src/engine/Preprocessor.cpp

```diff
@@ -781,6 +794,278 @@ void Preprocessor::dumpAllBounds( const String &message )
+        unsigned _neuron;
+    };
+
+    struct NeuronInformation
+    {
+        NeuronInformation()
+            : _weightedSumEquation( NULL )
+            , _activationFunction( NULL )
+        {
+        }
+
+        unsigned _weightedSumVariable;
+        unsigned _activationVariable;
+
+        PiecewiseLinearFunctionType _activationType;
+
+        const Equation *_weightedSumEquation;
+        const PiecewiseLinearConstraint *_activationFunction;
+    };
+
+    Map<Index, NeuronInformation> indexToNeuron;
+    Map<unsigned, Index> variableToIndex;
+    Map<unsigned, unsigned> layerToLayerSize;
+
+    ASSERT( !_preprocessed._networkLevelReasoner );
+
+    const List<Equation> &equations( _preprocessed.getEquations() );
+    const List<PiecewiseLinearConstraint *> plConstraints( _preprocessed.getPiecewiseLinearConstraints() );
+
+    // If the activation functions are not yet supported by the NLR,
+    // abort
+    for ( const auto constraint : plConstraints )
+        if ( !NetworkLevelReasoner::functionTypeSupported( constraint->getType() ) )
+            return false;
+
+    // Attempt to figure out the layered structure
+    List<unsigned> inputLayer;
+    List<unsigned> previousLayer;
+    List<unsigned> currentLayer;
+    List<unsigned> outputLayer;
+    unsigned currentLayerIndex = 1;
+
+    // The input and output layer should be defined in the input query
+    inputLayer = _preprocessed.getInputVariables();
+    previousLayer = inputLayer;
+    unsigned numVariablesHandledSoFar = inputLayer.size();
+    outputLayer = _preprocessed.getOutputVariables();
+    layerToLayerSize[0] = inputLayer.size();
+
+    // Store the input neurons in the topology
+    unsigned count = 0;
+    for ( const auto &neuron : inputLayer )
+    {
+        NeuronInformation neuronInfo;
+        neuronInfo._activationVariable = neuron;
+        Index index( 0, count );
+        indexToNeuron[index] = neuronInfo;
+        variableToIndex[neuron] = index;
+        ++count;
+    }
+
+    // Now, attempt to figure out the topology, layer by layer
+    while ( true )
+    {
+        currentLayer.clear();
+        for ( const auto &eq : equations )
+        {
+            // Only consider equalities
+            if ( eq._type != Equation::EQ )
+                continue;
+
+            Set<unsigned> eqVars = eq.getParticipatingVariables();
+            for ( const auto &var : previousLayer )
+                eqVars.erase( var );
+
+            if ( eqVars.size() == 1 )
+            {
+                // We have identified the weithed sum for a new neuron
+                NeuronInformation neuronInfo;
+                neuronInfo._weightedSumVariable = *eqVars.begin();
+                neuronInfo._weightedSumEquation = &eq;
+
+                Index index( currentLayerIndex, currentLayer.size() );
+                indexToNeuron[index] = neuronInfo;
+                currentLayer.append( *eqVars.begin() );
+
+                variableToIndex[*eqVars.begin()] = index;
```

Comment:
Maybe it makes sense from readability perspective to name *eqVars.begin(), or to use neuronInfo._weightedSumVariable in subsequent calls?

Reply:
Done